### ISSUE TITLE
p2p/addrbook: failed sanity check "Cant add old address to new bucket"

### ISSUE THREAD

**Issue by ebuchman** [tendermint/tendermint#1581]
Just saw the following on gaia-5001

```
I[05-16|12:43:07.097] Saving AddrBook to file                      module=p2p book=/root/.gaiad/config/addrbook.json size=74
I[05-16|12:43:07.100] Ensure peers                                 module=p2p numOutPeers=9 numInPeers=8 numDialing=0 numToDial=1
I[05-16|12:43:07.100] Will dial address                            module=p2p addr=4d927e3403668b22d751cd892fe7430d30a90ae7@108.28.150.189:46656
I[05-16|12:43:07.100] We need more addresses. Sending pexRequest to random peer module=p2p peer="Peer{MConn{35.184.182.143:46656} 30b45459e4881680c0ef1750fde136fefa6c3b98 out}"
I[05-16|12:43:07.100] Dialing peer                                 module=p2p address=4d927e3403668b22d751cd892fe7430d30a90ae7@108.28.150.189:46656
E[05-16|12:43:07.266] Failed Sanity Check! Cant add old address to new bucket module=p2p book=/root/.gaiad/config/addrbook.json ka="&{Addr:678503e6c8f50db7279c7da3cb9b072aac4bc0d5@35.193.188.125:46656 Src:b0dd378c3fbc4c156cd6d302a799f0d2e4227201@159.89.121.174:46656 Attempts:0 LastAttempt:2018-05-16 11:03:52.639711442 +0000 UTC LastSuccess:2018-05-16 11:03:52.639711442 +0000 UTC BucketType:2 Buckets:[37]}" bucket=192
E[05-16|12:43:07.266] Failed Sanity Check! Cant add old address to new bucket module=p2p book=/root/.gaiad/config/addrbook.json ka="&{Addr:30b45459e4881680c0ef1750fde136fefa6c3b98@35.184.182.143:46656 Src:b0dd378c3fbc4c156cd6d302a799f0d2e4227201@159.89.121.174:46656 Attempts:0 LastAttempt:2018-05-16 11:02:04.187007528 +0000 UTC LastSuccess:2018-05-16 11:02:04.187007528 +0000 UTC BucketType:2 Buckets:[56]}" bucket=192
```

**Comment by ebuchman** [comment]
It's possible we're seeing the same ID but a new IP

**Comment by ValarDragon** [comment]
Changing IP but not node ID doesn't cause this. (e.g. the following test passes)
```
func Test1581(t *testing.T) {
	fname := createTempFileName("addrbook_test")
	defer deleteTempFile(fname)

	addr, err := p2p.NewNetAddressString("678503e6c8f50db7279c7da3cb9b072aac4bc0d5@35.193.188.125:26656")
	require.Nil(t, err)
	src, err := p2p.NewNetAddressString("b0dd378c3fbc4c156cd6d302a799f0d2e4227201@159.89.121.174:26656")
	require.Nil(t, err)

	book := NewAddrBook(fname, true)
	book.SetLogger(log.TestingLogger())
	book.AddAddress(addr, src)
	book.MarkAttempt(addr)
	book.MarkGood(addr)

	// Double check that adding a peer again doesn't error
	err = book.AddAddress(addr, src)
	require.Nil(t, err)

	// Try changing ip but keeping the same node id. (35 -> 36)
	addr2, err := p2p.NewNetAddressString("678503e6c8f50db7279c7da3cb9b072aac4bc0d5@36.193.188.125:26656")
	require.Nil(t, err)
	err = book.AddAddress(addr2, src)
	require.Nil(t, err)
}
```

**Comment by ValarDragon** [comment]
Is the intended behavior for this that you can't update the IP for a node ID you already know about? If so, I think the fix is relatively straightforward, change https://github.com/tendermint/tendermint/blob/master/p2p/pex/addrbook.go#L666 to: `if ka.isOld() && (ka.Addr.ID == addr.ID) {`.

**Comment by ebuchman** [comment]
```
pex$ go test -run 15 -v
=== RUN   Test1581
--- PASS: Test1581 (0.00s)
PASS
ok      github.com/tendermint/tendermint/p2p/pex        0.016s
pex$ 
```

I'm running the code and not seeing the issue

**Comment by ebuchman** [comment]
I think you're right tho. The lookup is already by ID, so we should already have ka.Addr.ID == addr.ID. I think we only need to check `if ka.isOld()` and if so just ignore the new version with the updated IP address.

**Comment by ValarDragon** [comment]
Wow this is really weird. So the test doesn't fail like it is, but if I add a 
`fmt.Println("Hallo")` in `addToNewBucket` (line 529), then the test gives me the following:
```
=== RUN   TestUpdatingOldPeerIP
Hallo
Hallo
E[08-06|03:39:42.006] Failed Sanity Check! Cant add old address to new bucket ka="&{Addr:55550ab448cf58ec0d3cd21dc674e643da3bf301@218.174.95.153:16405 Src:a1190d57da6bcf0d04ffcdb49621ff929727c86e@75.18.65.236:41140 Attempts:0 LastAttempt:2018-08-05 22:39:42.006412076 -0500 CDT m=+0.007040149 LastSuccess:2018-08-05 22:39:42.006412076 -0500 CDT m=+0.007040149 BucketType:2 Buckets:[43]}" bucket=34
--- PASS: TestUpdatingOldPeerIP (0.00s)
PASS
ok  	github.com/tendermint/tendermint/p2p/pex	0.010s
```
I have no idea why this is the case.

**Comment by jackzampolin** [comment]
Closing as likely not an issue anymore. Please reopen if I'm wrong.

**Comment by ValarDragon** [comment]
I dont understand the reason for closing bugs without first confirming they are still present? I believe this was not fixed.

**Comment by ChatchaiCBV** [comment]
I found the same problem too. Anyone kindly advice. Thank you very much.

**Comment by njmurarka** [comment]
Just saw this:

```
Cant add old address to new bucket module=p2p
```

**Comment by fxfactorial** [comment]
I get this in the certik testnet - folks this is a real issue - no progress for 2 years?

**Comment by ValarDragon** [comment]
Using my old test, I've confirmed that it non-deterministically produces this error with high probability. (Its like 75% of the time, go test -v shows a warning, but oddly won't actually error)

After doing the fix I suggest in https://github.com/tendermint/tendermint/issues/1581#issuecomment-410554539, the error is fixed. I'll submit a PR for this

**Comment by ValarDragon** [comment]
Wow, I spent some time understanding whats going on here, and this sanity check saved us from a potentially serious p2p issue @ebuchman @melekes . Essentially if this sanity check wasn't here, any peer could overwrite the IP for every p2p node you're connected to, enabling an easy eclipse attack!

The problem is what we deduced. Moreover the test was correct, just it only worked probabilistically due to a probabilistic ignoring of peers. It never caused the test to fail, because the internal sanity check that reported the error didn't return an error, so it only logged the err.

The code I posted for generating the error and fixing it works! I'll submit a PR fixing & documenting this
